src/main/java/br/com/sunty/controller/course/CourseAdminController.java

```diff
@@ -1,55 +1,123 @@
+import br.com.sunty.models.category.dto.subcategory.AdminSubCategoryDto;
 import br.com.sunty.models.course.AdminCourseDto;
 import br.com.sunty.models.course.Course;
+import br.com.sunty.models.course.dto.AdminEditCourseForm;
+import br.com.sunty.models.course.dto.AdminEditCourseView;
+import br.com.sunty.models.course.dto.AdminNewCourseForm;
+import br.com.sunty.models.instructor.Instructor;
 import br.com.sunty.repository.course.CourseRepository;
+import br.com.sunty.repository.instructor.InstructorRepository;
 import br.com.sunty.repository.subcategory.SubCategoryRepository;
 import org.springframework.data.domain.Page;
 import org.springframework.data.domain.Pageable;
 import org.springframework.data.web.PageableDefault;
 import org.springframework.http.HttpStatus;
 import org.springframework.stereotype.Controller;
 import org.springframework.ui.Model;
+import org.springframework.validation.BindingResult;
 import org.springframework.web.bind.annotation.GetMapping;
 import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.PostMapping;
 import org.springframework.web.server.ResponseStatusException;
 
+import javax.validation.Valid;
+import java.util.List;
+
+import static org.springframework.http.HttpStatus.NOT_FOUND;
+
 @Controller
 public class CourseAdminController {
     private final CourseRepository courseRepository;
     private final SubCategoryRepository subCategoryRepository;
+    private final InstructorRepository instructorRepository;
 
-    public CourseAdminController(CourseRepository courseRepository, SubCategoryRepository subCategoryRepository) {
+    public CourseAdminController(CourseRepository courseRepository, SubCategoryRepository subCategoryRepository, InstructorRepository instructorRepository) {
         this.courseRepository = courseRepository;
         this.subCategoryRepository = subCategoryRepository;
+        this.instructorRepository = instructorRepository;
     }
 
 
     @GetMapping("/admin/courses/{categoryUrlCode}/{subcategoryUrlCode}")
     public String findAll(@PathVariable String subcategoryUrlCode,
+                          @PathVariable String categoryUrlCode,
                           @PageableDefault(size = 5)
                           Pageable pageable,
                           Model model) {
 
-        SubCategory subCategory = subCategoryRepository.findByUrlCode(subcategoryUrlCode)
-                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, subcategoryUrlCode));
+        AdminSubCategoryDto adminSubCategoryDto = subCategoryRepository.findByUrlCode(subcategoryUrlCode)
+                .map(AdminSubCategoryDto::new)
+                .orElseThrow(() -> new ResponseStatusException(NOT_FOUND, subcategoryUrlCode));
+
+        SubCategory subcategory = subCategoryRepository.findByUrlCode(subcategoryUrlCode)
+                .orElseThrow(() -> new ResponseStatusException(NOT_FOUND, subcategoryUrlCode));
```

Comment:
Não entendi o porquê de termos o `AdminSubCategoryDto` e a `SubCategory` aqui... :thinking:

Reply:
É por que eu usei o Derived Queries lá no course repository, pra fazer um findAllBySubCategory, dai o parametro que ele me solicita é uma subcategory. tentei passar o dto mas sem sucesso. E o subcategorydto pq nao to enviando nada de entidade inteira pro usuario. Ou como é uma listagem eu poderia enviar a entidade toda?

Reply:
Opa, entendi.

Se você observar bem, dá pra inverter as ordens das coisas e chamar o `subCategoryRepository` apenas 1 vez!

```suggestion
        SubCategory subcategory = subCategoryRepository.findByUrlCode(subcategoryUrlCode)
                .orElseThrow(() -> new ResponseStatusException(NOT_FOUND, subcategoryUrlCode));

                AdminSubCategoryDto adminSubCategoryDto = new AdminSubCategoryDto(subCategoryRepository);

```